internal/provider/resource_access_offer.go

```diff
@@ -0,0 +1,436 @@
+				},
+			},
+			// ID required for imports
+			"id": schema.StringAttribute{
+				Computed: true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"offer_url": schema.StringAttribute{
+				Description: "The url of the offer for access management. If this is changed the resource will be deleted and a new resource will be created.",
+				Required:    true,
+				Validators: []validator.String{
+					ValidatorMatchString(func(s string) bool {
+						_, err := crossmodel.ParseOfferURL(s)
+						return err == nil
+					}, "offer_url must be a valid offer string."),
+				},
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+		},
+	}
+}
+
+func (a *accessOfferResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	// Check first if the client is configured
+	if a.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "access offer", "create")
+		return
+	}
+	var plan accessOfferResourceOffer
+
+	// Read Terraform configuration from the request into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Get the users
+	var users []string
+	resp.Diagnostics.Append(plan.Users.ElementsAs(ctx, &users, false)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Get the offer
+	offerURLStr := plan.OfferURL.ValueString()
+	response, err := a.client.Offers.ReadOffer(&juju.ReadOfferInput{
+		OfferURL: offerURLStr,
+	})
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to create access offer resource, got error: %s", err))
+		return
+	}
+	a.trace(fmt.Sprintf("read offer %q at %q", response.Name, response.OfferURL))
+
+	accessStr := plan.Access.ValueString()
+	// Call Offers.GrantOffer
+	for _, user := range users {
+		err := a.client.Offers.GrantOffer(&juju.GrantRevokeOfferInput{
+			User:     user,
+			Access:   accessStr,
+			OfferURL: offerURLStr,
+		})
+		if err != nil {
+			resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to create access offer resource, got error: %s", err))
+			return
+		}
+	}
+	plan.ID = types.StringValue(newAccessOfferIDFrom(offerURLStr, accessStr, users))
+
+	// Set the plan onto the Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+}
+
+func (a *accessOfferResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	// Check first if the client is configured
+	if a.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "access offer", "read")
+		return
+	}
+	var state accessOfferResourceOffer
+
+	// Get the Terraform state
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Get information from ID
+	offerURL, _, _ := retrieveAccessOfferDataFromID(ctx, state.ID, state.Users, &resp.Diagnostics)
```

Comment:
why ignore the access level from offer ID. i think you should use it to filter out users that don't have the matching access level.

Reply:
Implementation has changed but the provider still gets all users from the Offer, creates a map and add them to the state.